### ISSUE TITLE
Adding a BufferedInputStream

### ISSUE THREAD

**Issue by dwnusbaum** [StanzaOrg/lbstanza-old#36]
On the Stanza mailing list, it was mentioned that it would be nice to have a `BufferedInputStream` that wraps an existing `InputStream` and buffers up to `n` characters from that stream. I think this would be great! I have come up with the following, and would like some feedback on the desired interface.

**Type:**
`deftype BufferedInputStream <: InputStream`

**Creation:**
Create a stream with a buffer of size `n`:
`public defn BufferedInputStream(i:InputStream, n:Int) -> BufferedInputStream`

Create a stream with the default buffer size:
`public defn BufferedInputStream(i:InputStream) -> BufferedInputStream`

**Interface:**
1. Like Stanza's `StringInputStream`: 
   
   Identical to the `peek?` multi for `StringInputStream`
   `public defmulti peek? (s:BufferedInputStream, i:Int) -> Char|False`
   `public defmulti peek? (s:BufferedInputStream) -> Char|False`
   
   Like `fill`, but does not remove the read characters from the stream:
   `public defmulti peek-fill (c:CharArray, r:Range, s:BufferedInputStream) -> Int`
2. Similar to Java's `BufferedInputStream`, where a stream can be reset to a marked position:
   
   Marks the stream's current position, allowing the stream to be reset to the current position if no more than n characters are read from the stream after the call to mark. Returns false if the mark could not be set, otherwise true:
   `public defmulti mark (s:BufferedInputStream, n:Int) -> False|True`
   
   Resets the stream to the most recent mark. Returns false if there is no valid mark set, otherwise true:
   `public defmulti reset (s:BufferedInputStream) -> False|True`
3. Something else: Explain it in a comment!

Which of the options do you prefer, and how would you change it? I lean slightly towards option 1, but would like some feedback before submitting a pull request.

**Comment by CuppoJava** [comment]
That looks fantastic. I am also leaning towards option 1. The marking and resetting looks odd to me and needlessly imperative and stateful. 

Question: with `peek-fill`, does it error out if you try to read more than `n` characters? Or does it fulfill the read by silently throwing away the buffer?

**Comment by dwnusbaum** [comment]
I agree about marking and resetting.

I think it would make the most sense to throw an error if you try to read more than `n` characters, which would make `n` both the buffer size and the lookahead limit for the `peek` functions.

**Comment by CuppoJava** [comment]
That sounds perfectly good to me. If you submit a pull request I'll integrate it into the 0.10.0 that I'm working on releasing.

**Comment by dwnusbaum** [comment]
Great! How can I test my changes?

I tried using `scripts/make-compiler.stanza` to compile a test compiler which I would then have used to test the `BufferedInputStream`, but it failed with a linker error.

I also tried compiling packages for `core.stanza` and `collections.stanza`.  However, when I compiled my test against those packages using `-pkg-path`, the compilation failed with a linker error again.

Both linker errors were:

```
Undefined symbols for architecture x86_64:
  "_file_read_block", referenced from:
      __L1089521 in tmp1-8bd1fd.o
  "_file_seek", referenced from:
      __L1030727 in tmp1-8bd1fd.o
  "_file_set_length", referenced from:
      __L1030712 in tmp1-8bd1fd.o
  "_file_skip", referenced from:
      __L1030742 in tmp1-8bd1fd.o
  "_file_write_block", referenced from:
      __L1089687 in tmp1-8bd1fd.o
  "_get_file_size", referenced from:
      __L1030699 in tmp1-8bd1fd.o
```

Seems like I'm not picking up the C bindings correctly? I'm on OS X, and gcc links to clang on my system, if either of those matters.

**Comment by CuppoJava** [comment]
Hi Devin.

The easiest way to try out your changes is to try compiling the `core` and `collections` library together with your test code. 

Let's assume that your new changes are in `mycore.stanza`, and your test code is in `mytests.stanza`. Then you can compile your test code using:

```
stanza mycore.stanza stanzadir/src/collections.stanza mytests.stanza -o mytests -ccfiles stanzadir/runtime/driver.c
```

Let me know if that works.
  -Patrick

**Comment by dwnusbaum** [comment]
Thanks for the help. I moved my `core.stanza` and `test.stanza` into a new directory, and then ran the following:

```
stanza core.stanza ../lbstanza/core/collections.stanza test.stanza -o test -ccfiles ../lbstanza/runtime/driver.c
```

Which gave me this error:

```
duplicate symbol _main in:
    /var/folders/3t/m1k0ycjj1_j3wxf1934m_55w0000gn/T/driver-05ee62.o
    /var/folders/3t/m1k0ycjj1_j3wxf1934m_55w0000gn/T/driver-a728e3.o
duplicate symbol _get_stdout in:
    /var/folders/3t/m1k0ycjj1_j3wxf1934m_55w0000gn/T/driver-05ee62.o
    /var/folders/3t/m1k0ycjj1_j3wxf1934m_55w0000gn/T/driver-a728e3.o
duplicate symbol _get_stderr in:
    /var/folders/3t/m1k0ycjj1_j3wxf1934m_55w0000gn/T/driver-05ee62.o
    /var/folders/3t/m1k0ycjj1_j3wxf1934m_55w0000gn/T/driver-a728e3.o
duplicate symbol _get_stdin in:
    /var/folders/3t/m1k0ycjj1_j3wxf1934m_55w0000gn/T/driver-05ee62.o
    /var/folders/3t/m1k0ycjj1_j3wxf1934m_55w0000gn/T/driver-a728e3.o
duplicate symbol _get_eof in:
    /var/folders/3t/m1k0ycjj1_j3wxf1934m_55w0000gn/T/driver-05ee62.o
    /var/folders/3t/m1k0ycjj1_j3wxf1934m_55w0000gn/T/driver-a728e3.o
duplicate symbol _get_errno in:
    /var/folders/3t/m1k0ycjj1_j3wxf1934m_55w0000gn/T/driver-05ee62.o
    /var/folders/3t/m1k0ycjj1_j3wxf1934m_55w0000gn/T/driver-a728e3.o
duplicate symbol _current_time_us in:
    /var/folders/3t/m1k0ycjj1_j3wxf1934m_55w0000gn/T/driver-05ee62.o
    /var/folders/3t/m1k0ycjj1_j3wxf1934m_55w0000gn/T/driver-a728e3.o
duplicate symbol _current_time_ms in:
    /var/folders/3t/m1k0ycjj1_j3wxf1934m_55w0000gn/T/driver-05ee62.o
    /var/folders/3t/m1k0ycjj1_j3wxf1934m_55w0000gn/T/driver-a728e3.o
duplicate symbol _resolve_path in:
    /var/folders/3t/m1k0ycjj1_j3wxf1934m_55w0000gn/T/driver-05ee62.o
    /var/folders/3t/m1k0ycjj1_j3wxf1934m_55w0000gn/T/driver-a728e3.o
ld: 9 duplicate symbols for architecture x86_64
clang: error: linker command failed with exit code 1 (use -v to see invocation)
```

I tried removing the `#ifdef PLATFORM_WINDOWS` blocks in driver.c, but that didn't help. Any ideas?

**Comment by CuppoJava** [comment]
Okay. I'm seeing the same thing. Give me a sec.

**Comment by CuppoJava** [comment]
Ah. I was being silly. You don't need the `-ccfiles` flag at all.

```
stanza core.stanza ../lbstanza/core/collections.stanza test.stanza -o test
```

**Comment by dwnusbaum** [comment]
Hmm, that is what I was trying at first, and it still gives me the "Undefined symbols ..." error.  Do I need to clear out the packages folders or anything?

Here is my test code:

```
defpackage test:
    import core

defn test ():
    val i = StringInputStream("Hello, world!")
    val s = BufferedInputStream(i)
    while peek?(s) is Char :
        println(get-char(s))

test()
```

**Comment by CuppoJava** [comment]
Hmm. It seems like its pointing to an obsolete Stanza version.

Here's three things to verify:
1. Use `stanza version` to verify that you're using 0.9.6. 
2. Type `head ~/.stanza` and double check that `install-dir` is pointing to the correct Stanza installation directory.
3. Open up `stanzadir/runtime/driver.c` and verify that `file_read_block` is indeed defined. It is supposed to be.

Let me know what happens. Thanks for being patient.
  -Patrick

**Comment by dwnusbaum** [comment]
That was it! I had pulled a fresh copy down to make my changes, but the copy on my path was 0.9.4. It's working now, thanks for the help!

**Comment by CuppoJava** [comment]
Great! And thanks for the report. I'll put some thought towards whether malformed system installations can be detected more gracefully instead of spitting out this incomprehensible linker error.

**Comment by dwnusbaum** [comment]
Closed by #37
